### ISSUE TITLE
Test Case:  Metrics

### ISSUE THREAD

**Issue by leobardoramon** [gosha1128/NYUAMS#556]
IP: http://52.0.50.223/js/login.html
login: leo
pw: l6shooter
domain: Local

Metrics data does not completely load

**Comment by backprops** [comment]
OK I fixed it.  Remember, if you look at the metrics page - at the top left ( right under the weekly_digest_activity tab ) there is a progress counter.  The page should interactively update as it progresses.

**Comment by leobardoramon** [comment]
Weekly seems to work.

So far, "studio" and "university" are take 15 minutes to load -which is probably okay- but the progress bar seems to get stuck at "Processing transations for metrics..."
![studio](https://cloud.githubusercontent.com/assets/2165796/12836743/80c490d4-cb89-11e5-8b5f-8cc4590fa404.png)
![university](https://cloud.githubusercontent.com/assets/2165796/12836742/80c4a8b2-cb89-11e5-8672-4ae4dd9a115c.png)
![weekly](https://cloud.githubusercontent.com/assets/2165796/12836741/80c21f02-cb89-11e5-80c1-26034e009e52.png)

**Comment by gosha1128** [comment]
Ok carve some time Sunday ... Get me login details...

**Comment by leobardoramon** [comment]
See login details in issue body

**Comment by gosha1128** [comment]
So I understand - the issue right now is it gets stuck at processing transactions, right ?

**Comment by gosha1128** [comment]
OK, I looked at it.  

Now that there is more data, i see a few issues that will be best resolved by updating the SQL schema.  We've been holding off this but it seems necessary now.

In order to do this i'll need a few things so that we aren't doing this on the production instance - which can still be used while we work on the dev instance:

1) copy the existing data into a new "dev" SQL instance for development

2) bring up a new "dev" EC2 instance for development and point it to the SQL instance.

3) i'll work on the required updates on dev and we will test on this instance.  

4) once its approved - then we will apply the SQL updates to the production instance over a weekend or planned down time.

This might take several days to a week or so.

**Comment by leobardoramon** [comment]
Sounds Good.

This is directly related to #493

Andrew and Shelly will be pleased to know that this will be the next maintenance issue.

**Comment by gosha1128** [comment]
OK, let me know when the dev instances are ready.  As I go I will outline here in github the schema changes.  Please find all github issues around schema changes / performance issues and reopen.  They all relate to schema updates.

**Comment by leobardoramon** [comment]
Also note that Shelly cannot access weekly metrics reports - normally they get emailed every Sunday.

How long has this been happening?

George, is there a way that one of us could quickly create a Metric pdf with less than 24 hours of notice? -this may happen

**Comment by NYU-AMS** [comment]
I approve of this plan.

-Andrew

**Comment by leobardoramon** [comment]
ec2 instance
<img width="2560" alt="screen shot 2016-03-02 at 3 30 45 pm" src="https://cloud.githubusercontent.com/assets/2165796/13474416/f0fbdc92-e08b-11e5-90b8-8ce6649bd868.png">
<img width="2560" alt="screen shot 2016-03-02 at 3 30 47 pm" src="https://cloud.githubusercontent.com/assets/2165796/13474415/f0faf7dc-e08b-11e5-830b-e5f79574d790.png">

rds instance
<img width="1623" alt="screen shot 2016-03-02 at 3 32 43 pm" src="https://cloud.githubusercontent.com/assets/2165796/13474465/19cb752e-e08c-11e5-8059-737c34cf8ca9.png">
<img width="1630" alt="screen shot 2016-03-02 at 3 32 53 pm" src="https://cloud.githubusercontent.com/assets/2165796/13474464/19cac69c-e08c-11e5-9cc0-201a5cf116e9.png">
<img width="1626" alt="screen shot 2016-03-02 at 3 33 01 pm" src="https://cloud.githubusercontent.com/assets/2165796/13474466/19cd83dc-e08c-11e5-945c-dc63977fcef8.png">

**Comment by backprops** [comment]
Great!

So, you'll have to login to the dev ec2 instance and change the SQL
connection string to the new RDS instance and then start EBS.

Then, you should verify that any changes you are making at the dev ec2 is
*only* changing the data at the dev RDS and not production RDS.

That way we know we can treat this as a playground for the schema updates
that is not updating production what so ever...

George

**Comment by leobardoramon** [comment]
I've added the "dev3" ec2 instance's ip address to the inbound rules of the "dev1" rds instance's security group

<img width="1629" alt="screen shot 2016-03-02 at 3 37 49 pm" src="https://cloud.githubusercontent.com/assets/2165796/13475153/482c38f6-e08f-11e5-9841-38369fecd635.png">
<img width="1637" alt="screen shot 2016-03-02 at 3 54 36 pm" src="https://cloud.githubusercontent.com/assets/2165796/13475154/482f0b30-e08f-11e5-9958-5c26ac76092c.png">

How do I change the connections string to the new RDS instance?  I'm assuming this done via Terminal.

**Comment by gosha1128** [comment]
Yes, if you you ever bring up a cloned ec2 instance (i.e., ebs is already
installed ), you need to do a few things:

1) login via ssh

2) continue as root

>> sudo bash

3) locate ebs installation

>> cd /opt/NYUAMS

4) pull all the latest code changes

>> git fetch

4) make sure you are on the right "branch"

for production:

>> git checkout master

for this maintenance branch, lets call it "dev_maintenance1"

>> git checkout dev_maintenance1

[ NOTE:  I have not created this branch yet so it will likely complain, but
you should know the process when it is created. ]

5) make sure the SQL connection string points to the right server

you need to open and edit the file db/config.js

notice the sql connection string and sql login parameters

6) then restart ebs ( either manual or background mode )

manual debug mode:  this is useful to see warnings and errors bringing up
the server

>> node app.js nfs

background server mode:  this is for when you are happy and just want to
run the server forever

>> forever app.js start nfs

7) after ebs has started (either manual or background), you will be able to
access the web front-end.  login and make sure the data you see represents
the SQL instance you have set.

**Comment by leobardoramon** [comment]
Great, the EBS launched :)

Followed your instructions, see 6-minute video here:
https://drive.google.com/open?id=0B_9JEK6SVk1bZGttOHlGdVZONHM

dev1 rds instance activity does not connect to PROD servers or rds instance what so ever :)

I created a project on dev3 ec2 instance server ( http://54.164.124.142/ ) and then check lgsdb.nyu.edu to see if it would appear...it did NOT :)

See 1.5 minute video here:
https://drive.google.com/open?id=0B_9JEK6SVk1bck1qUzR6UktLNjA

**Comment by backprops** [comment]
Nice !!

**Comment by leobardoramon** [comment]
See new DEV system credentials in issue #562

**Comment by gosha1128** [comment]
OK, so I tracked down the slowest call at the database ( get_order
information for an order ).  You don't typically notice it in other areas
of the app, but in metrics you will because it needs to make this call for
every order.

I will be spending a bit of time optimizing this call at the database, so
please don't depend on the DEV system being up for a few days...

**Comment by leobardoramon** [comment]
> comment from @gosha1128 copied from issue #562 :
>
>Another thing...metrics works the way it does so that you always get the latest numbers when you go
>there. We could also do something where we just "cache" the last computation of metrics so that
>when you go to metrics it shows the last status of metrics. Maybe we put a button with lets you
>explicity recompute now. This will be easy....let me know what you guys want.

1. Yes to "refresh" button for metrics page.

a few questions:

1. Can we have the system pull latest metrics data every day at midnight?
   - so that that will be the data that is shown when the page is being requested

2. Can we have the system cache metrics data for previous months?
   - this way the system is only recalling new data

**Comment by backprops** [comment]
Yes on #1.  But that means we need to re-introduce the scheduler and digest
features.  So, make sure to make those issues a priority.

Also, there is currently no user interface to browse past metrics.  The
easy thing to do here is the following:  metrics reports are exported in a
"metrics" directory a the file server - the file name has a date and
timestamp.

George

**Comment by gosha1128** [comment]
Also, i'm wondering if its best to have one EC2 instance dedicated just to
metrics and scheduler stuff.

That way, it doesn't interfere with common use of EBS.

We can have a special way you start the server, like this:

forever start app.js nfs metrics

That server instance handles automatic metrics etc...

**Comment by leobardoramon** [comment]
George, I like this idea. It seems very efficient and clean.  How many components will this involve?

**Comment by gosha1128** [comment]
Anything that involved stuff being "scheduled"

create metrics report schedule

send metrics email schedules

send digest email schedules

**Comment by backprops** [comment]
OK, can you create an issue around this ?

Call it "implement EBS scheduler mode"

**Comment by gosha1128** [comment]
You need to create a new EC2 instance in DEV for this for me.  It can be a clone of the primary EBS server if you wish...
